src/main/java/pl/edu/icm/pl/mxrdr/extension/workflow/XdsSupplyMissingValuesStep.java

```diff
@@ -0,0 +1,137 @@
+import pl.edu.icm.pl.mxrdr.extension.workflow.utils.FileContentReplacer;
+import pl.edu.icm.pl.mxrdr.extension.workflow.utils.FileContentReplacer.PatternAndReplacement;
+
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class XdsSupplyMissingValuesStep extends FilesystemAccessingWorkflowStep {
+
+    static final String STEP_ID = "xds-supply-missing-values";
+
+    private static final Set<String> SOURCE_FIELDS_SET = Initializer.initializeSourceFieldsSet();
+    private static final Map<String, PatternAndReplacement> PATTERNS = Initializer.initializePatternMap();
+
+    private DatasetVersionServiceBean versionService;
+
+    public XdsSupplyMissingValuesStep(Map<String, String> inputParams, DatasetVersionServiceBean versionService) {
+        super(inputParams);
+        this.versionService = versionService;
+    }
+
+    @Override
+    protected WorkflowStepResult.Source runInternal(WorkflowExecutionContext workflowExecutionContext, Path path) throws Exception {
+        List<PatternAndReplacement> actions = prepareReplacementActions(workflowExecutionContext);
+        FileContentReplacer.of(XdsAdjustResultStep.XDS_INP, path)
+                .add(actions)
+                .replace();
+        return Success::new;
+    }
+
+    @Override
+    public WorkflowStepResult resume(WorkflowExecutionContext workflowExecutionContext, Map<String, String> map, String s) {
+        throw new UnsupportedOperationException("This step does not pause");
+    }
+
+    @Override
+    public void rollback(WorkflowExecutionContext workflowExecutionContext, Failure failure) { }
+
+    // -------------------- PRIVATE --------------------
+
+    private List<PatternAndReplacement> prepareReplacementActions(WorkflowExecutionContext workflowExecutionContext) {
+        Map<String, String> replacementValues = readReplacementValues(workflowExecutionContext);
+        return PATTERNS.keySet().stream()
+                .filter(k -> replacementValues.get(k) != null)
+                .map(k -> {
+                    PatternAndReplacement patterns = PATTERNS.get(k);
+                    return FileContentReplacer.PatternAndReplacement.of(patterns.pattern,
+                            String.format(patterns.replacement, replacementValues.get(k)));
+                })
+                .collect(Collectors.toList());
+    }
+
+    private Map<String, String> readReplacementValues(WorkflowExecutionContext executionContext) {
+        return findVersion(executionContext)
+                .map(DatasetVersion::getDatasetFields)
+                .orElse(Collections.emptyList()).stream()
+                .filter(f -> MxrdrMetadataField.DATA_COLLECTION.getValue()
+                        .equals(extractFieldName(f)))
+                .findFirst()
```

Comment:
Feature envy here - above 5 lines could be replaced by `Optional<DatasetField> getDatasetField(MxrdrMetadataField)` method in `DatasetVersion` class with added bonus to readability.

Reply:
I was considering using that method, but it was written in not very efficient way and using method that now is deprecated, so I've chosen to do it in alternative way.

Reply:
I wasn't even aware that such method already existed there. The more reason to use it. If it's implementation is inefficient, improve it. It's the worst kind of duplication and hurting maintainability badly to scatter the same logic with different implementations in places where it does not belong.